File: src/voice_terminal/stopword.py
"""Stop word detector — detects 'thank you' (or custom) at end of transcript."""

import re
from dataclasses import dataclass


@dataclass
class DetectionResult:
    detected: bool
    command: str


class StopWordDetector:
    """Detects a stop phrase at the end of a transcript.

    Handles variations like ". thank you", ", thank you", "! thank you".
    """

    def __init__(self, stop_word: str = "thank you") -> None:
        self.stop_word = stop_word
        self._stop_lower = stop_word.lower()

    def process(self, transcript: str) -> DetectionResult:
        """Check if transcript ends with stop word.

        Returns DetectionResult with detected=True and the clean command
        (stop word and trailing punctuation stripped).
        """
        trimmed = transcript.strip()
        lower = trimmed.lower()

        # Direct match at end
        if lower.endswith(self._stop_lower):
            command = trimmed[: len(trimmed) - len(self.stop_word)].strip()
            command = re.sub(r"[,.\s]+$", "", command)
            return DetectionResult(detected=True, command=command)

        # Variations with punctuation before stop word
        for sep in [". ", ", ", "! "]:
            variation = sep + self._stop_lower
            if lower.endswith(variation):
                command = trimmed[: len(trimmed) - len(variation)].strip()
                return DetectionResult(detected=True, command=command)

        return DetectionResult(detected=False, command="")

File: ui/stopword.js
/**
 * Stop word detector — detects "thank you" (or custom) at end of transcript.
 *
 * Normalizes text before matching: removes punctuation, collapses spaces,
 * lowercases. So "Thank You", "thank.you", "THANK  YOU!" all match.
 */

class StopWordDetector {
  constructor(stopWord = "thank you") {
    this.stopWord = stopWord;
  }

  /**
   * Normalize text for comparison:
   * - Remove punctuation
   * - Collapse multiple spaces
   * - Trim
   * - Lowercase
   */
  _normalize(text) {
    return text
      .replace(/[.,!?;:'"()\[\]{}]/g, "")
      .replace(/\s+/g, " ")
      .trim()
      .toLowerCase();
  }

  /**
   * Check if transcript ends with stop word (after normalization).
   * Returns { detected, command } where command has stop word stripped.
   */
  process(transcript) {
    const normalizedTranscript = this._normalize(transcript);
    const normalizedStopWord = this._normalize(this.stopWord);

    if (!normalizedTranscript.endsWith(normalizedStopWord)) {
      return { detected: false, command: "" };
    }

    // Map normalized stop-word position back to original transcript
    const stopWordStart =
      normalizedTranscript.length - normalizedStopWord.length;

    const trimmed = transcript.trim();
    let origIndex = 0;
    let normCount = 0;

    for (let i = 0; i < trimmed.length && normCount < stopWordStart; i++) {
      const char = trimmed[i];
      // Only count chars that survive normalization
      if (!/[.,!?;:'"()\[\]{}]/.test(char)) {
        if (char === " " || /\s/.test(char)) {
          // Only count if not collapsing multiple spaces
          if (i === 0 || !/\s/.test(trimmed[i - 1])) {
            normCount++;
          }
        } else {
          normCount++;
        }
      }
      origIndex = i + 1;
    }

    const command = trimmed.substring(0, origIndex).trim();
    return { detected: true, command };
  }
}
